docs: add 4.3 release notes (#2728)
canonical/charmcraft: docs/release-notes/charmcraft-4.3.rst

Author: https://canonical-ubuntu-documentation-library--2728.com.readthedocs.build/charmcraft/2728/release-notes/charmcraft-4.3/

Charmcraft 4.3 release notes
=============================

23 June 2026

Learn about the new features, changes, and fixes introduced in Charmcraft 4.3.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`__. It
comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.3 brings the following new features.

Ubuntu 26.04 LTS support for the reactive plugin
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :ref:`reactive plugin <reference-charmcraft-yaml-reactive-plugin>` now supports Ubuntu 26.04 LTS (Resolute Raccoon)
as a base. This allows you to build reactive charms targeting the latest Ubuntu LTS
release. The plugin also now adds the ``universe`` APT repository automatically, as

> Python 3 virtual environments require the python3-venv package found there.

Reviewer: ```suggestion
The :ref:`reactive plugin <reference-charmcraft-yaml-reactive-plugin>` can now build
charms on the Ubuntu 26.04 LTS base. The plugin also now adds the ``universe`` APT
repository automatically, as Python 3 virtual environments require the python3-venv
package found there.
```

> This allows you to build reactive charms targeting the latest Ubuntu LTS
release.

Tried to combine this with the previous sentence. I appreciate the separation between
the change to the mechanics and what that change allows users to do, but it felt like
there was a lot of redundancy.

> (Resolute Raccoon)

Are you particularly attached to the codename? I find it to be noisy in most cases, and the [style guide](https://docs.ubuntu.com/styleguide/en#branding) states that it's optional.

> The plugin also now adds the ``universe`` APT repository, ...

This sentence doesn't make it clear to me how this relates to the 26.04 base.

Author: I'm not strongly attached to the codename, but I do often ctrl+f the codename before the version number.

Author: ```suggestion
The :ref:`reactive plugin <reference-charmcraft-yaml-reactive-plugin>` can now build
charms on the Ubuntu 26.04 LTS base. The plugin also now adds the ``universe`` APT
repository automatically, as Python 3 virtual environments require the python3-venv
package, which is in the ``universe`` repository on Ubuntu 26.04 LTS.
```

> See :doc:`/howto/migrate-bases/change-to-ubuntu-2604` for guidance on migrating your

Reviewer: ```suggestion
See :ref:`howto-change-to-ubuntu-26-04` for guidance on migrating your
```

charm to Ubuntu 26.04 LTS.

Strict validation of part names

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Part naming constraints
~~~~~~~~~~~~~~~~~~~~~~~
```

Part names specified in ``charmcraft.yaml`` under ``parts`` are now validated more
strictly and cannot contain forward slashes (``/``) when targeting the ``ubuntu@26.04``

> base or higher.

Reviewer: ```suggestion
Part names are now validated more
strictly and cannot contain forward slashes (``/``) in charms built on the ``ubuntu@26.04``
base or higher.
```
> Part names specified in ``charmcraft.yaml`` under ``parts``

This phrasing makes me think that there can be part names specified elsewhere. If you want to keep the original specificity, I'd rather go with something like:

```suggestion
Part names, declared with the ``parts`` key in the ``charmcraft.yaml`` file, are now
validated more strictly. They can't contain forward slashes (``/``) in charms built on
the ``ubuntu@26.04`` base or higher.
```

> Valkey support in 12-factor app init templates
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Disregard if the feature existed prior to its recent inclusion in the init template.
```suggestion
Valkey integration for 12-factor app charms
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
From the description, it sounds like the interface is the new feature, and the changes to the init template are how the user interacts with it.

Author: That new feature is in the PAAS charms library. The only change to Charmcraft is updating the init templates.

The init templates for 12-factor app frameworks (Django, Flask, FastAPI, Express.js,
Go, and Spring Boot) now include a commented-out ``valkey_client`` relation. If you
declare this optional interface in your charm's ``requires``, Charmcraft automatically

> injects the necessary dpcharmlibs-interfaces charm Python package.

Reviewer: Similar to my comment on the title for this section, let's focus on the Valkey integration itself if that is indeed the new feature. Something like:

```
You can integrate Valkey into your 12-factor app charm by declaring the
``valkey_client`` relation in the charm's ``requires`` key. This relation is now included
in the comments of any charm projects initialized with a 12-factor app profile.

If this relation is declared, Charmcraft injects the necessary dpcharmlibs-interfaces
Python package.
```

Again, disregard if it's just a change to the init template.

The following environment variables are made available to your app when the

> ``valkey_client`` relation is active:

Reviewer: ```suggestion
If the ``valkey_client`` relation is declared, your can make use of the following
environment variables in your app:
```
"are made available" sounds a little passive

- ``VALKEY_HOSTNAME``
- ``VALKEY_PORT``
- ``VALKEY_USERNAME``
- ``VALKEY_PASSWORD``

See :ref:`extensions` for details.

Updated charm init templates

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Improved test scaffolding
~~~~~~~~~~~~~~~~~~~~~~~~~
```
Lots of things manifest as updates to the init templates, as evidenced by the previous section. We're underselling it by not emphasizing the substance of those changes! 😃

New charms created with ``charmcraft init`` benefit from several improvements to the

> generated project templates:

Reviewer: ```suggestion
The ``charmcraft init`` command now generates charm project files with improved
test scaffolding:
```

- Templates now use `pytest-jubilant
  <https://pytest-jubilant.readthedocs.io/en/stable/>`__ as the test runner for
  integration tests, replacing the previous custom helpers. The
  ``@pytest.mark.juju_setup`` marker is included in integration test scaffolding for
  machine and Kubernetes profiles.
- Integration test templates for machine and Kubernetes profiles have improved comments
  pointing to the canonical guide on writing integration tests.

COS directory structure validation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When packing a 12-factor app charm, Charmcraft now validates the structure of any
custom COS (Canonical Observability Stack) directory. If the directory layout is
incorrect, Charmcraft reports an error at pack time rather than letting an invalid

> structure be discovered only at runtime.

Reviewer: ```suggestion
incorrect, Charmcraft reports an error when packing instead of letting an invalid
structure cause errors at runtime.
```
This one is pretty subjective. I just don't like terms like "build time" or "pack time" in writing.

> Documentation improvements

Reviewer: ```suggestion

Documentation improvements
```
This is the least important of these comments, but most of the headings above this one are preceded by only one blank line, while the following heads have two. Keeping this consistent is a nice bit of source hygiene.

Reviewer 2: Something we could lint for, eventually...

.. _release-4.3:

Charmcraft 4.3 release notes
=============================

23 June 2026

Learn about the new features, changes, and fixes introduced in Charmcraft 4.3.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`__. It
comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.3 brings the following new features.

Ubuntu 26.04 LTS support for the reactive plugin
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :ref:`reactive plugin <reference-charmcraft-yaml-reactive-plugin>` can now build
charms on the Ubuntu 26.04 LTS base. The plugin also now adds the ``universe`` APT
repository automatically, as Python 3 virtual environments require the python3-venv

> package, which is in the ``universe`` repository on Ubuntu 26.04 LTS.

Reviewer: Optional
```suggestion
charms on the Ubuntu 26.04 LTS base. Because the python3-venv package was moved
to the `universe` APT repository on Ubuntu 26.04 LTS, the plugin now adds this
repository automatically.
```
Just a little more concise.

See :ref:`howto-change-to-ubuntu-26-04` for guidance on migrating your
charm to Ubuntu 26.04 LTS.

Part naming constraints
~~~~~~~~~~~~~~~~~~~~~~~

Part names are now validated more
strictly and cannot contain forward slashes (``/``) in charms built on the ``ubuntu@26.04``
base or higher.

Valkey support in 12-factor app init templates
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The init templates for 12-factor app frameworks (Django, Flask, FastAPI, Express.js,
Go, and Spring Boot) now include a commented-out ``valkey_client`` relation. If you
declare this optional interface in your charm's ``requires``, Charmcraft automatically
injects the necessary dpcharmlibs-interfaces charm Python package.

> If the ``valkey_client`` relation is declared, your can make use of the following

Reviewer: ```suggestion
If the ``valkey_client`` relation is declared, you can make use of the following
```
My bad
